### PR TITLE
Avoid endless loop on EventDrivenTaskSource

### PR DESCRIPTION
**Pull request by losipiuk**
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by dekimir — COMMENTED** [review]
Would it be possible to have a test that fails without this change?

**Comment by brybacki** [comment]
I suppose a unit test for taskSource process can be done. But it probably needs to replicate the ideas from the main loop of Scheduler, so it is not a nice black box test.

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenTaskSource.java [hunk 1]
```diff
@@ -125,18 +127,22 @@ class EventDrivenTaskSource
         this.getSplitTimeRecorder = requireNonNull(getSplitTimeRecorder, "getSplitTimeRecorder is null");
     }
 
-    public synchronized ListenableFuture<AssignmentResult> process()
+    public synchronized Optional<ListenableFuture<AssignmentResult>> process()
     {
         checkState(!closed, "closed");
         checkState(future == null || future.isDone(), "still in process");
 
+        if (finished) {
```
Comment (brybacki) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenTaskSource.java:135]: Nice, a stop condition.

**Comment by losipiuk** [comment]
> Would it be possible to have a test that fails without this change?

Added an assertion that `EventDrivenTaskSource` is finished (returns `Optional.empty()` as soon as we get finished split assignments for all the sources.

**Review by dekimir — APPROVED** [review]

**Comment by losipiuk** [comment]
CI: https://github.com/trinodb/trino/pull/21424/files

**Pull request merged**